contracts/Smt.metasol

```diff
@@ -0,0 +1,155 @@
+        require(depthDifference > 0, "depthDifference must be non-zero");
+        uint8 oldDepth = tree.depth;
+        // underflow protection done by solidity 0.8
+        uint8 newDepth = oldDepth - depthDifference;
+
+        for (uint8 level = newDepth; level < oldDepth; level = inc(level)) {
+            require(tree.nodes[level][1] == EMPTY_SUBTREE, "Subtree must be empty");
+            // If we will delete this:
+            // delete tree[level+1][0];
+            // delete emptySubtreeCache[level+1];
+            // refund will be smaller than gas spent :(
+        }
+
+        tree.depth = newDepth;
+    }
+
+    function getRoot(MerkleTree storage tree) internal view returns (bytes32) {
+        if (tree.nodes[tree.depth][0] == EMPTY_SUBTREE) {
+            return tree.emptySubtreeCache[tree.depth];
+        }
+        return tree.nodes[tree.depth][0];
+    }
+
+    function _calculateEmptySubtreeHash(MerkleTree storage tree, uint8 toLevel) internal {
+        uint8 currentDepth = tree.depth;
+        bytes32 prev = tree.emptySubtreeCache[currentDepth];
+
+        for (uint8 level = currentDepth+1; level <= toLevel; level = inc(level)) {
+            // We write the hash to both memory and storage at the same time
+            // in order to use it in the next iteration. It is cheaper to read
+            // from memory instead of storage, so we save it here.
+            tree.emptySubtreeCache[level] = prev = {{hash}}(abi.encodePacked(prev, prev));
+        }
+    }
+
+    function _updateNode(MerkleTree storage tree, uint8 level, uint i) internal {
+        // level MUST be > 0
+        // level MUST be < 2^{depth}
+
+        bytes32 v0;
+        bytes32 v1;
+
+        unchecked {
+            uint i0 = 2*i;
+            uint i1 = 2*i+1;
+
+            v0 = tree.nodes[level-1][i0];
+            v1 = tree.nodes[level-1][i1];
+        }
+
+        if ((v0 == EMPTY_SUBTREE) && (v1 == EMPTY_SUBTREE)) {
+            delete tree.nodes[level][i];
+            // If we will use:
+            // tree[level][i] = EMPTY_SUBTREE;
+            // We will spend much more gas
+            return;
+        }
+
+        if (v0 == EMPTY_SUBTREE) {
+            unchecked { v0 = tree.emptySubtreeCache[level-1]; }
+        }
+
+        if (v1 == EMPTY_SUBTREE) {
+            unchecked { v1 = tree.emptySubtreeCache[level-1]; }
+        }
+
+        tree.nodes[level][i] = {{hash}}(abi.encodePacked(v0, v1));
+    }
+
+    function _modifyTree(MerkleTree storage tree, uint index, bytes32 hashedElement) internal {
+        require(tree.depth > 0, "Tree must be initialized");
+        require(index < totalElements(tree), "Index out of bounds");
+        tree.nodes[0][index] = hashedElement;
+        for (uint8 level = 1; level <= tree.depth; level = inc(level)) {
+            // index >> level == index / 2**level
+            uint currentIndex = index >> level;
+            tree._updateNode(level, currentIndex);
+        }
```

Comment:
Small improvement in gas consumption by changing the for loop to as follows:

uint currentIndex = index;
for (uint8 level = 1; level <= tree.depth; level = inc(level)) {
            // index >> level == index / 2**level
            currentIndex = currentIndex >> 1;
            tree._updateNode(level, currentIndex);
        }